File: solutions/golang/547/solution.go
var graph map[int][]int
var seen []bool

func findCircleNum(isConnected [][]int) int {
	N := len(isConnected)
	graph = make(map[int][]int)
	seen = make([]bool, N)
	for i := 0; i < N; i++ {
		for j := i + 1; j < N; j++ {
			if isConnected[i][j] == 1 {
				graph[i] = append(graph[i], j)
				graph[j] = append(graph[j], i)
			}
		}
	}

	ans := 0
	for i := 0; i < N; i++ {
		if seen[i] == true {
			continue
		}
		seen[i] = true
		dfs(i)
		ans++
	}
	return ans
}

func dfs(node int) {
	for _, neighbor := range graph[node] {
		if seen[neighbor] == true {
			continue
		}
		seen[neighbor] = true
		dfs(neighbor)
	}
}

File: solutions/cpp/547/solution.cpp
class Solution {
public:
    unordered_map<int, vector<int>> graph;
    vector<bool> seen;
    int N;
    int findCircleNum(vector<vector<int>>& isConnected) {
        N = isConnected.size();
        seen = vector(N, false);
        // Create unidirect graph
        for (int i = 0; i < N; i++) {
            for (int j = i + 1; j < N; j++) {
                if (isConnected[i][j]) {
                    graph[i].push_back(j);
                    graph[j].push_back(i);
                }
            }
        }

        int ans = 0;
        for (int i = 0; i < N; i++) {
            if (seen[i] == true) {
                continue;
            }
            seen[i] = true;
            dfs(i);
            ans++;
        }
        return ans;
    }

    void dfs(int node) {
        for (int neighbor : graph[node]) {
            if (seen[neighbor] == true) {
                continue;
            }
            seen[neighbor] = true;
            dfs(neighbor);
        }
    }
};